jl4-core/src/L4/Evaluate.hs

```diff
@@ -632,6 +655,43 @@ runLit :: Lit -> Eval Value
 runLit (NumericLit _ann num) = pure (ValNumber num)
 runLit (StringLit _ann str)  = pure (ValString str)
 
+runBuiltin :: Stack -> [Value] -> UnaryBuiltinFun -> Eval Value
+runBuiltin s vals = \case
+  UnaryIsInteger -> do
+    val <- expect1Number s vals
+    pure $ valBool $ isJust $ isInteger val
+  UnaryRound -> do
+    val <- expect1Number s vals
+    pure $ ValNumber $ toRational $ round @Rational @Integer val
+  UnaryCeiling -> do
+    val <- expect1Number s vals
+    pure $ ValNumber $ toRational $ ceiling @Rational @Integer val
+  UnaryFloor -> do
+    val <- expect1Number s vals
+    pure $ ValNumber $ toRational $ floor @Rational @Integer val
+
+expect1 :: Stack -> [a] -> Eval a
+expect1 s = \case
+  [x] -> pure x
+  xs ->
+    exception (RuntimeTypeError $ "Expected 1 argument but got " <> Text.show (length xs)) s
+
+expectNumber :: Stack -> Value -> Eval Rational
+expectNumber s = \case
+  ValNumber f -> pure f
+  _ -> exception (RuntimeTypeError "Expected rational number.") s
+
+expect1Number :: Stack -> [Value] -> Eval Rational
+expect1Number s vs = do
+  v <- expect1 s vs
+  expectNumber s v
+
+expectInteger :: Stack -> Rational -> Eval Integer
+expectInteger stack n = do
+  case isInteger n of
+    Nothing -> exception (NotAnInteger n) stack
+    Just i -> pure i
```

Comment:
it would be good if the error message included which operation the error stemmed from, in `expectNumber`, it would be good, if the error message would include the "but got" part of the sentence.

Reply:
This is an internal error, the typechecker should already catch the case of `expect1Number`.